sprocs/grading_jobs_process_external.sql

```diff
@@ -57,28 +58,43 @@ BEGIN
         grading_received_at = received_time,
         grading_started_at = start_time,
         grading_finished_at = finish_time,
+        gradable = new_gradable,
         score = grading_jobs_process_external.score,
         correct = (grading_jobs_process_external.score >= 1.0),
         feedback = grading_jobs_process_external.feedback
     WHERE id = grading_job_id
     RETURNING *
     INTO grading_job;
 
-    UPDATE submissions
-    SET
-        graded_at = grading_job.graded_at,
-        score = grading_job.score,
-        correct = grading_job.correct,
-        feedback = grading_job.feedback
-    WHERE id = grading_job.submission_id;
+    IF new_gradable = FALSE THEN
+        UPDATE submissions
+        SET
+            gradable = FALSE,
+            feedback = grading_jobs_process_external.feedback
```

Comment:
Should we be setting `format_errors` rather than `feedback` here?

I'm thinking that the rule is that "valid" submissions must have `format_errors = {}` and may have non-empty `feedback`, while "invalid" submissions must have `feedback = {}` and may have non-empty `format_errors`.

Does this seem reasonable? It would fit with the big flowchart at the bottom of https://prairielearn.readthedocs.io/en/latest/dev-guide/

Perhaps we could update that flowchart by adding `submission.format_errors = <value>` to the "Ungradable submission" ellipse?

Or are we setting `feedback` here because external graders are putting compile errors into feedback?

Reply:
Yep, external grader results contain everything in `feedback`, they don't have a `format_errors` key.

Reply:
Perhaps we should add `format_errors` as an allowed key in `results.json`? I'm thinking that the rule will basically be that external graders work just like internal graders in terms of which variables they can access and which ones they can modify?

Reply:
Sure, I'd be on board with that.  I'm thinking if there are any `format_errors` they would be displayed between the "Score" and "Test Results" sections (I think this is where any messages or output would currently be displayed anyways)

![image](https://user-images.githubusercontent.com/1790491/78576380-34cf1900-77f2-11ea-8cd5-79d892296c99.png)

Reply:
What do we want our exact rules to be for `gradable`, `score`, and `format_errors`?

I think at the moment we [only allow](https://github.com/PrairieLearn/PrairieLearn/blob/master/question-servers/freeform.js#L376) a `server.py` `grade()` function to set `format_errors`, and then we [enforce](https://github.com/PrairieLearn/PrairieLearn/blob/master/question-servers/freeform.js#L1166) the rule that if `format_errors` is not empty then `gradable = false`. We don't enforce `score = null` in this case, but maybe we should? Also maybe we should force `partial_scores` to `{}` to avoid leaking any information?

I'm thinking about something like:

* After both `parse()` and `grade()`, if we have `format_errors` then set `gradable = false`, `score = null`, and `partial_scores = {}`.
* If we don't have `format_errors` then set `gradable = true`.

The above rules would apply to both internal and external grading.

Thoughts?

Reply:
Yeah, I like `score = null` and `partial_scores = {}`, it's better than having some random garbage in there.

I also do like having an explicit `gradable` flag for the grader, though I guess for consistency with the internal grading scheme we should just get that from `format_errors`.

Reply:
I'm also fine with having `gradable` as an explicit flag. Would we add it to the internal grade() as well, to keep consistency?

Reply:
I would think so, that might be best to do in a separate PR as I could see many of the python elements needing to be updated.

Reply:
![image](https://user-images.githubusercontent.com/1790491/78690342-ecc8f880-78bc-11ea-97f1-4dc132939c39.png)

Thoughts? Unrelated, but I think we should not display "Test results" if there are none.

Reply:
![image](https://user-images.githubusercontent.com/1790491/78690966-b6d84400-78bd-11ea-9684-24a2c6a6c1d4.png)

Reply:
@mwest1066 should we be actually setting `partial_scores` to null here?  I see that's what the value is when `parse()` fails.

Reply:
I think we should set `partial_scores` to `null` and also `score` to `null` (or 0? whatever it would be if we detected a format error during parse). Otherwise we risk leaking grading information. For example, say there is a pl-number-input that already graded successfully. We don't want to show the student that their answer to that was correct.

I think we should also set `partial_scores` to `null` (and `score` to whatever) here: https://github.com/PrairieLearn/PrairieLearn/blob/master/question-servers/freeform.js#L1179

Reply:
Yeah, this is what I have on the diagram I think:

https://raw.githubusercontent.com/PrairieLearn/PrairieLearn/7009d4a7e26738e07fc48f11e96bb878c659e94a/docs/question-flow.png

Reply:
Perfect! Now we just need the code to match the diagram 😄

Reply:
Oh, that should be in there already!  Internal grading jobs will set score, partials, and correct to `null`, external will set `score` and `correct` to `null` since the partials don't ever have a value anyway I think?

Reply:
Hahaha, ok, I guess I need to look at the actual code 😄 

I think that it might be a good idea to set `partial_scores = null` as well, just in case we change things in the future to run both the internal and external grader (I actually think this could be a good idea).